File: backend/src/core/machineId.ts
/**
 * 机器码生成/验证模块
 * 简化版本，只保留生成和验证逻辑，移除系统操作
 */

import crypto from 'crypto'

/**
 * 生成随机机器码 (UUID 格式)
 * 格式: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
 */
export function generateMachineId(): string {
  return crypto.randomUUID().toLowerCase()
}

/**
 * 验证机器码格式
 * 支持两种格式:
 * - UUID 格式: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
 * - 纯32位十六进制 (Linux machine-id 格式)
 */
export function isValidMachineId(machineId: string): boolean {
  if (!machineId || typeof machineId !== 'string') {
    return false
  }

  // UUID 格式: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
  const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i

  // 纯32位十六进制 (Linux machine-id 格式)
  const hexRegex = /^[0-9a-f]{32}$/i

  return uuidRegex.test(machineId) || hexRegex.test(machineId)
}

/**
 * 将32位十六进制转换为 UUID 格式
 */
export function formatAsUUID(hex: string): string {
  const clean = hex.replace(/-/g, '').toLowerCase()
  if (clean.length !== 32) {
    return clean
  }
  return `${clean.slice(0, 8)}-${clean.slice(8, 12)}-${clean.slice(12, 16)}-${clean.slice(16, 20)}-${clean.slice(20)}`
}

/**
 * 将 UUID 格式转换为32位十六进制
 */
export function formatAsHex(uuid: string): string {
  return uuid.replace(/-/g, '').toLowerCase()
}

/**
 * 标准化机器码格式（统一为 UUID 格式）
 */
export function normalizeMachineId(machineId: string): string {
  if (!machineId) {
    return generateMachineId()
  }

  const clean = machineId.replace(/-/g, '').toLowerCase()

  // 如果是32位十六进制，转换为 UUID 格式
  if (clean.length === 32 && /^[0-9a-f]+$/i.test(clean)) {
    return formatAsUUID(clean)
  }

  // 如果已经是 UUID 格式，直接返回小写
  if (isValidMachineId(machineId)) {
    return machineId.toLowerCase()
  }

  // 无效格式，生成新的
  return generateMachineId()
}

File: packages/gateway/core/machine_id.go
// Package core contains the core business logic for the kiro-gateway,
// including proxy orchestration, session management, and pricing calculations.
package core

import (
	"regexp"
	"strings"

	"github.com/google/uuid"
)

// uuidRegex matches a standard UUID format (lowercase hex with dashes).
var uuidRegex = regexp.MustCompile(`^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`)

// hexRegex matches a 32-character lowercase hex string (UUID without dashes).
var hexRegex = regexp.MustCompile(`^[0-9a-f]{32}$`)

// GenerateMachineID creates a new random machine ID in UUID v4 format.
func GenerateMachineID() string {
	return uuid.New().String()
}

// IsValidMachineID checks whether the given string is a valid machine ID.
// Accepts both standard UUID format (with dashes) and 32-character hex format.
func IsValidMachineID(id string) bool {
	lower := strings.ToLower(id)
	return uuidRegex.MatchString(lower) || hexRegex.MatchString(lower)
}

// FormatAsUUID converts a 32-character hex string into standard UUID format
// with dashes (8-4-4-4-12). If the input is not exactly 32 hex characters,
// it is returned as-is.
func FormatAsUUID(hex string) string {
	lower := strings.ToLower(hex)
	if !hexRegex.MatchString(lower) {
		return hex
	}
	return lower[:8] + "-" + lower[8:12] + "-" + lower[12:16] + "-" + lower[16:20] + "-" + lower[20:]
}

// FormatAsHex converts a UUID string into a 32-character hex string by
// removing all dashes. Non-UUID inputs are returned with dashes stripped.
func FormatAsHex(id string) string {
	return strings.ReplaceAll(strings.ToLower(id), "-", "")
}

// NormalizeMachineID lowercases and validates the given machine ID.
// If the ID is invalid or empty, a new machine ID is generated automatically.
func NormalizeMachineID(id string) string {
	trimmed := strings.TrimSpace(id)
	if trimmed == "" {
		return GenerateMachineID()
	}

	lower := strings.ToLower(trimmed)

	// Already a valid UUID.
	if uuidRegex.MatchString(lower) {
		return lower
	}

	// Valid 32-char hex, convert to UUID format.
	if hexRegex.MatchString(lower) {
		return FormatAsUUID(lower)
	}

	// Invalid format, generate a new one.
	return GenerateMachineID()
}
